### ISSUE TITLE
 lazy load app style

### ISSUE THREAD

**Issue by HemTech** [single-spa/single-spa-angular#187]
I’m trying to create separate bundle of my style sheet 
I’m implementing multiple themes and want to load css file runtime base on user request. I have implemented it with normal angular 8 application and able generate separated bundle for each theme (ng build)  so can load the require theme at run time. Below is my angular.json configuration. 

> "build": {
>           "builder": "@angular-devkit/build-angular:browser",
>           "options": {
>             "outputPath": "dist/app1 ",
>             "index": "src/index.html",
>             "main": "src/main.ts",
>             "polyfills": "src/polyfills.ts",
>             "tsConfig": "src/tsconfig.app.json",
>            
> ],
>             "styles": [
>               "src/styles.scss",
>               {
>                 "input": "src/styles/themes/light-theme.scss",
>                 "lazy": true,
>                 "bundleName": "light-theme"
>               },
>               {
>                 "input": "src/styles/themes/dark-theme.scss",
>                 "lazy": true,
>                 "bundleName": "dark-theme"
>               }
>             ],
>         
> 

We are using @angular-builders/custom-webpack:browser builder for single spa app and it is not generating individual bundle for css file.

**Comment by filoxo** [comment]
Are you reporting a bug?

**Comment by HemTech** [comment]
I think the feature is missing in custom-webpack then we can consider as issue.

**Comment by arturovt** [comment]
Could you provide a _minimal reproducible example_?

**Comment by arturovt** [comment]
Any progress on creating a reproducible example?

**Comment by HemTech** [comment]
I'm working on a release this week so will share an example by Saturday.

**Comment by arturovt** [comment]
That's great! Looking forward.

**Comment by HemTech** [comment]
The minimal rep here: https://github.com/HemTech/Lazyloading

I have created application having portal (normal angular 8 app) and Header(single spa app) projects. In angular.Json, I have added configuration to bundle themes into two separate files.

**Portal:**
"styles": [
              "projects/portal/src/styles.scss",
              {
                "input": "projects/portal/src/styles/themes/light-theme.scss",
                "lazy": true,
                "bundleName": "light-theme"
              },
              {
                "input": "projects/portal/src/styles/themes/dark-theme.scss",
                "lazy": true,
                "bundleName": "dark-theme"
              }
            ]
And 
"architect": {
        "build": {
          "builder": "@angular-devkit/build-angular:browser",
      

**Header**:
 "styles": [
              "projects/header/src/header-styles.scss",
              {
                "input": "projects/header/src/styles/themes/light-theme.scss",
                "lazy": true,
                "bundleName": "header-light-theme"
              },
              {
                "input": "projects/header/src/styles/themes/dark-theme.scss",
                "lazy": true,
                "bundleName": "header-dark-theme"
              }
            ]

And 
 "architect": {
        "build": {
          "builder": "@angular-builders/custom-webpack:browser",

When we run build command, Header is not creating individual css file as per the configuration.

**Comment by arturovt** [comment]
I just cloned your repo, ran `ng build --project header` and Webpack generated separate files for `lazy` CSS:

![image](https://user-images.githubusercontent.com/7337691/83289915-724a9400-a1ee-11ea-94d8-ee7f5d2c243e.png)

**Comment by arturovt** [comment]
And I actually wouldn't load styles like that:
1) you have to inject `script` element and set its `src` to `http://APP/header-light-theme.js`
2) each `.js` bundle with CSS inside also contains `css-loader/runtime` and `style-loader`

You've got a dynamic import and you can inject styles w/o having to specify implicit paths:
```ts
export const enum Theme {
  Dark,
  Light,
}

@Injectable({ providedIn: 'root' })
export class ThemeResolver {
  private resolver: Observable<unknown>;

  async loadSpecificTheme(theme: Theme) {
    if (theme === Theme.Dark) {
      this.resolver = defer(() => import('style-loader!./dark-theme.scss'));
    } else if (theme === Theme.Light) {
      this.resolver = defer(() => import('style-loader!./light-theme.scss'))
    }

    await this.resolver.toPromise();
  }
}
```

P.S. this is just a pseudo-code :point_up: 

And somewhere else in the app:
```ts
constructor(themeResolver: ThemeResolver) {
  themeResolver.loadSpecificTheme(Theme.Dark);
}
```

**Comment by HemTech** [comment]
Thank you so much! @arturovt - I will use style-loader.

If you use ng build –configuration=production –project=header then it will not generate theme.css files. 
In angualr.json - we have an entry for customwebconfig to use Single-spa-angular/lib/webpack.

Below code of our singe-spa-webpack removed the css and merge in to main bundle.

// @ts-ignore
  const mergedConfig = webpackMerge.smart(config, singleSpaConfig);
  removePluginByName(mergedConfig.plugins, "IndexHtmlWebpackPlugin");
  removeMiniCssExtract(mergedConfig);
  if (Array.isArray(mergedConfig.entry.styles)) {
    // We want the global styles to be part of the "main" entry. The order of strings in this array
    // matters -- only the last item in the array will have its exports become the exports for the entire
    // webpack bundle
    mergedConfig.entry.main = [...mergedConfig.entry.styles, ...mergedConfig.entry.main];
  }
  // Remove bundles
  delete mergedConfig.entry.polyfills;
  delete mergedConfig.entry.styles;
  delete mergedConfig.optimization.runtimeChunk;
  delete mergedConfig.optimization.splitChunks;
  return mergedConfig;
};

**Comment by arturovt** [comment]
Yes, you're right. I forgot to build with `--prod` flag.

Styles are not emitted in separate files because `MiniCssExtractPlugin` is removed from Webpack config.

I'm just wondering if it has to be removed completely.

We've got such code:
https://github.com/single-spa/single-spa-angular/blob/389a16cb186ed5fd975199ff2e4670084c47a84f/src/webpack/index.ts#L73

Which replaces `mini-css-extract-plugin` loader with `style-loader`. Thus the main CSS file (`styles.scss`) gets bundled into JS file and then `style-loader/runtime` injects this CSS into DOM.

And when we have such `styles`:
```json
[
  "src/styles.scss",
  "src/cool-bro.scss",
  {
    "input": "src/lazy.scss",
    "lazy": true,
    "bundleName": "super-lazy"
  }
]
```

Then Angular CLI adds another `entry` to Webpack config, thus we will have:
```js
const config = {
  entry: {
    styles: ['/src/styles.scss', 'src/cool-bro.scss'],
    'super-lazy': ['/src/lazy.scss']
  }
}
```

And Webpack's `module.rules` will contain rule with pre-configured `include` statement:
```js
rule = {
  test: /\.scss$|\.sass$/,
  include: [
    'src/styles.scss',
    'src/cool-bro.scss',
    'src/lazy.scss'
  ],
  use: [
    'mini-css-extract-plugin',
    '@angular-devkit/build-angular/raw-css-loader',
    'postcss-loader',
    'sass-loader'
  ]
}
```

In conclusion, Webpack will apply `mini-css-extract-plugin` for `styles.scss + files included into config via single path (e.g. src/cool-bro.scss)` and for all files that are `lazy: true` (configured via an object in config).

The current algorithm might be a solution to this problem:

```{r, tidy=FALSE, eval=FALSE, highlight=FALSE }

`rule` = find rule that contains `mini-css-extract-loader`

loop through each `rule.include`
  loop through each `webpackConfig.entry`
    check if `rule.include[i]` is a part of `entry.styles`
      if `rule.include[i]` IS then
        create rule that will apply `style-loader` for that file

    check if `rule.include[i]` is a part of `entry[KEY]`
      if `rule.include[i]` IS then
        create rule that will apply `mini-css-extract-loader` for that file

```

So basically we will not remove `mini-css-extract-plugin` loader and its plugin completely. We will just distribute files and apply different loaders to them:
```
[src/styles.scss, src/cool-bro.scss] -> style-loader (they will get into main.js)
[src/lazy.scss] -> mini-css-extract-plugin (will be emitted as a separate file)
```

@joeldenning wtyd?

**Comment by friendly-tech** [comment]
any update on this - we also would like to load a theme/css file at runtime depending on the customer.
I have tried reversing some of the webpack shenanigans done by `singleSpaAngularWebpack` but unable to understand whats needed.

**Comment by joeldenning** [comment]
I am not planning on looking at this issue anytime soon. If others would like to jump in to help out, that would be great.

**Comment by orosro** [comment]
I am working on a PR for adding an extra option when calling the function to generate the new webpack config: https://github.com/orosro/single-spa-angular/tree/issue-187-lazy-load-app-style

It fits for my needs, the one where I don't have any style file injected by the angular app, but I let the portal config handle the style injection/removal.

So in this case, all the styles would be rendered if they had lazy set to true. I think a better solution would be indeed to apply the style-loader for normal styles and keep the miniCssExtract plugin for the lazy ones

Any feedback on the solution would be appreciated. I will try implementing the pseudo code from above later

**Comment by arturovt** [comment]
I've tested it with the latest Angular 13 and it seems like this issue is not reproducible anymore. The lazy `.scss` file is extracted into a JS file which then can be loaded through `System.import()` or just by loading it through HTTP. I've added an integration test for that case: https://github.com/single-spa/single-spa-angular/commit/217cdd5116dbc214e9abdb3007adac70c9cf2d60#diff-8172ff72e908e21a8b6b1b3d34f2ae36c9d33c31a291de7e10542407defd8fa5R14-R30.
